CONTRACTS: Add enumerative loop invariant synthesizer

qinheping opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.


Implement the functionality described below.

# Motivation

This loop invariant synthesizer use the idea counter-example-guided synthesis (CEGIS) to synthesize loop invariants for programs with only checks instrumented by `goto-instrument` with flag `--pointer-check`.

This PR contain the driver of the synthesizer and the verifier we use to check invariant candidates.

# Verifier
The verifier take as input a goto program with pointer checks and a map from loop id to loop invariant candidates. It first annotate and apply the loop invariant candidates into the goto model; and then simulate the CBMC api to verify the instrumented goto program. If there are some violations---loop invariants are not inductive, or some pointer checks fail---, it record valuation from trace generated by the back end to construct a formatted counterexample `cext`.

## Counterexample
A counterexample `cext` record valuations of variables in the trace, and other information about the violation. The valuation we record including
1. set of live variables upon the entry of the loop. 
2. the havoced value of all primitive-typed variables;
3. the havoced offset and the object size of all pointer-typed variables;
4. history values of 2 and 3.

The valuations will be used as true positive (history values) and true negative (havcoed valuation) to filter out bad invariant clause with the idea of the Daikon invariant detector in a following PR. However, in this PR we only construct the valuation but not actually use them.

# Synthesizer
Loop invariants we synthesize are of the form
`` (in_clause || !guard) && (!guard -> pos_clause)``
where `in_clause` and `out_clause` are predicates we store in two different map, and `guard` is the loop guard. The idea is that we separately synthesize the condition `in_clause` that should hold before the execution of the loop body, and condition `pos_clause` that should hold as post-condition of the loop.

When the violation happen in the loop, we update `in_clause`. When the violation happen after the loop, we update `pos_clause`. When the invariant candidate it not inductive, we enumerate strengthening clause to make it inductive.

To be more efficient, we choose different synthesis strategy for different type of violation
* For out-of-boundary violation, we choose to use the violated predicate as the new clause, which is the WLP of the violation if the violation is only dependent on the havocing instruction. **TODO**: to make it more complete, we need to implement a WLP algorithm
* For null-pointer violation, we choose `__CPROVER_same_object(ptr, __CPROVER_loop_entry(ptr))` as the new clause. That is, the havoced pointer should points to the same object at the start of every iteration of the loop. It is a heuristic choice. This can be extended with the idea of alias analysis if needed.
* For invariant-not-preserved violation, we enumerate strengthening clauses and check that if the invariant will be inductive after strengthening (disjunction with the new clause).

The synthesizer works as follow
1. initialize the two invariant maps,
2. verify the current candidates built from the two maps,
_a. return the candidate if there is no violation
_b. synthesize a new clause to resolve the **first** violation and add it to the correct map,
3. repeat 2.

The flag `synthesize-loop-invariants` will also apply synthesized loop contracts.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented:
You have an extensive PR description, which is really nice. Would you mind also including that in the commit message, or at least an executive summary thereof?

tautschnig commented on regression/contracts/loop_contract_synthesis_02/test.desc:
```diff
@@ -0,0 +1,11 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test shows that loop invariant with form of range predicates can be correctly
+synthesized for programs with only pointer checks but no other assertions.
```
So what happens when there _are_ additional assertions?

qinheping replied:
The synthesizer tries to solve violation one by one. If there are other violations beyond pointer checks and invariant checks, it will throw an exception saying the type of violation is unsupported. My plan is focusing on pointer checks in this PR, and create another PR to add the enumerating-and-check mode that keeps enumerating until all checks pass for other type of violations.

tautschnig commented on src/goto-programs/loop_ids.h:
```diff
@@ -31,6 +31,15 @@ struct loop_idt
   {
   }
 
+  loop_idt() : function_id(""), loop_number(-1)
+  {
+  }
+
+  loop_idt(const loop_idt &other)
+    : function_id(other.function_id), loop_number(other.loop_number)
+  {
+  }
+
```
Which piece of code needs these? At least the first one looks a bit dangerous and I wonder whether we might instead need to go for a bigger change where `loop_number` is made `optionalt<unsigned int>`.

qinheping replied:
When computing cause loop, we want a value to indicate that no cause loop found, which means that the violation is nothing about loop invariants. Therefore, synthesizing loop invariant doesn't help in such case.
Also, now `loop_number` is a member variable for every instruction. However, it make no sense to have a loop number for an instruction not in a loop. So I agree that changing `loop_number` from `unsigned int` to `optionalt<unsigned int>` is a good idea.
I will make the change for `loop_idt` in this PR. And then change other `loop_number` (those in `goto_programt::instructiont`) with another PR. What do you think?

kroening replied:
Do return an `optionalt<loop_idt>` when looking for a cause.
Please don't change `goto_programt::instructiont`. It's deliberate that the `loop_number` in the instruction is undefined unless the instruction forms a loop.

tautschnig commented on src/goto-instrument/synthesizer/synthesizer_utils.h:
```diff
@@ -50,7 +53,29 @@ get_loop_head(const unsigned int loop_number, goto_functiont &function);
+  const loop_idt &loop_id,
+  const goto_functiont &function,
+  unsigned location_number_of_target);
+
+/// Extract size of dynamic object from `expr`.
+std::size_t extract_size(const exprt &expr);
+
+/// Extract offset of dynamic object from `expr`.
+std::size_t extract_offset(const exprt &expr);
+
+/// convert hex string to size_t.
+std::size_t hex_to_size_t(const std::string &hex_str);
```
I don't think you're using these anywhere? If you need them for some follow-up work: please move them to this follow-up work and then we'll discuss again (they need to be implemented differently, but I'd like to better understand how they are used so that I can give the most appropriate advice).

tautschnig commented on src/goto-instrument/synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,482 @@
+          // entered the loop.
+          if(!entered_loop)
+          {
+            loop_entry_offsets[step.full_lhs] = extract_offset(
+              annotated_pointer_constant_expr->symbolic_pointer());
+          }
+
+          // Store the offset into the the map for havoced offset if this step
+          // is a loop havocing instruction.
+          if(is_loop_havoc(step.pc))
+          {
+            havoced_pointer_offsets[step.full_lhs] = extract_offset(
+              annotated_pointer_constant_expr->symbolic_pointer());
```
Please use `pointer_arithmetict` instead of your `extract_offset` implementation, unless this doesn't work for you.

tautschnig commented on src/goto-instrument/synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,482 @@
+            expr_try_dynamic_cast<annotated_pointer_constant_exprt>(
+              step.full_lhs_value);
+
+          object_sizes[step.full_lhs] =
+            extract_size(annotated_pointer_constant_expr->symbolic_pointer());
```
Use `pointer_arithmetict` to pick up the pointer, and do `pointer_offset_size(to_pointer_type(pointer.type().base_type(), ns))`; you'll then want to use `numeric_cast<std::size_t>` of the result.

tautschnig commented on src/goto-instrument/synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,482 @@
+
+bool cegis_verifiert::verify()
+{
+#if DEBUG
+#else
+  null_message_handlert null_message_handler;
+  ui_message_handlert ui_null_message_handler(null_message_handler);
+  messaget null_log(ui_null_message_handler);
+  log = null_log;
+#endif
```
This looks suspicious to me. We have `.debug()`, which should be debug-only output.

tautschnig commented on src/goto-instrument/CMakeLists.txt:
```diff
@@ -13,6 +13,7 @@ target_link_libraries(goto-instrument-lib
     linking
     big-int
     goto-programs
+    goto-checker
```
Should the synthesizer become a front-end of its own?

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -294,6 +294,12 @@ void code_contractst::check_apply_loop_contracts(
   havoc_assigns_targetst havoc_gen(to_havoc, ns);
   havoc_gen.append_full_havoc_code(loop_head_location, pre_loop_head_instrs);
+  // Add loop number to the havoc_code
+  Forall_goto_program_instructions(havoc_it, pre_loop_head_instrs)
```
I know we have these macros, but we should really be able to use ranged `for` expressions nowadays.

tautschnig commented on src/goto-instrument/havoc_utils.cpp:
```diff
@@ -16,7 +16,14 @@ Date: July 2021
 #include <util/std_code.h>
 
-#include <goto-programs/goto_program.h>
+bool is_loop_havoc(const goto_programt::const_targett &instruction)
```
Pass a `const goto_programt::instructiont` instead of an iterator.

tautschnig commented on src/goto-instrument/havoc_utils.cpp:
```diff
@@ -52,6 +59,8 @@ void havoc_utilst::append_object_havoc_code_for_expr(
   havoc.add_to_operands(expr);
   dest.add(goto_programt::make_other(havoc, location));
+  dest.instructions.back().source_location_nonconst().set_comment(
+    "Loop havocing instrumented by appling loop contracts");
```
If you really need to use a string comment to communicate this information, then please at least put this string in a macro. (And perhaps fix the typo: s/appling/applying/)

tautschnig commented on src/goto-instrument/synthesizer/enumerative_loop_invariant_synthesizer.cpp:
```diff
@@ -11,8 +11,62 @@ Author: Qinheping Hu
+      // __CPROVER_pointer_offset(v),
+      // __CPROVER_pointer_offset(__CPROVER_loop_entry(v))
+      // into the result.
+      result.push_back(unary_exprt(ID_pointer_offset, e, size_type()));
+      result.push_back(unary_exprt(
+        ID_pointer_offset,
+        unary_exprt(ID_loop_entry, e, e.type()),
+        size_type()));
+    }
```
We have `pointer_offset_exprt`.

tautschnig commented on src/goto-instrument/synthesizer/enumerative_loop_invariant_synthesizer.cpp:
```diff
@@ -32,19 +86,295 @@ void enumerative_loop_invariant_synthesizert::init_candidates()
+    {
+      // tmp_post variables are symbol lhs of ASSIGN.
+      if(
+        instruction.type() != goto_program_instruction_typet::ASSIGN ||
```
Use `instruction.is_assign()`

tautschnig commented on src/goto-instrument/synthesizer/enumerative_loop_invariant_synthesizer.cpp:
```diff
@@ -32,19 +86,295 @@ void enumerative_loop_invariant_synthesizert::init_candidates()
+}
+
+// Recursively extract symbols from `expr`
+void extract_symbol_exprt_rec(const exprt &expr, std::set<exprt> &result)
+{
+  if(expr.id() == ID_symbol)
+  {
+    result.insert(expr);
+  }
+  else
+  {
+    for(exprt::operandst::const_iterator it = (expr).operands().begin();
+        it != (expr).operands().end();
+        ++it)
+    {
+      extract_symbol_exprt_rec(*it, result);
+    }
+  }
+}
```
Does some function from `util/find_symbols.{h,cpp}` do the trick for you?

qinheping replied:
Yes, thank you!

tautschnig commented on src/goto-instrument/synthesizer/synthesizer_utils.cpp:
```diff
@@ -112,3 +117,96 @@ void annotate_invariants(
+}
+
+invariant_mapt combine_in_and_post_invariant_clauses(
+  invariant_mapt in_clauses,
+  invariant_mapt post_clauses,
+  invariant_mapt neg_guards)
```
Should all of these be const references?

qinheping replied:
Yes, fixed.

tautschnig reviewed (commented):
My apologies, my comments are a bit all-over-the-place and very surface level: My biggest concern is that we're building what could (should?) be a free-standing tool into `goto-instrument`. I really think that this deserves to be a tool of its own. Maybe you really want to copy https://github.com/diffblue/cbmc/pull/6526, which should be the bare-bones for a new tool.

qinheping commented:
> My apologies, my comments are a bit all-over-the-place and very surface level: My biggest concern is that we're building what could (should?) be a free-standing tool into `goto-instrument`. I really think that this deserves to be a tool of its own. Maybe you really want to copy #6526, which should be the bare-bones for a new tool.

It sounds great to me. I will create another PR to initialize the new tool `goto-synthesizer`. At the same time, let's keep pushing this PR to be merged. Once the PR for the new tool is approved, I will migrate the loop-contract synthesizer to the new tool.

qinheping committed:
Add enumerative loop invariant synthesizer

Implement the functionality described below.

Motivation
---
This loop invariant synthesizer use the idea counter-example-guided synthesis (CEGIS) to synthesize loop invariants for programs with only checks instrumented by `goto-instrument` with flag `--pointer-check`.

This PR contain the driver of the synthesizer and the verifier we use to check invariant candidates.

Verifier
---
The verifier take as input a goto program with pointer checks and a map from loop id to loop invariant candidates. It first annotate and apply the loop invariant candidates into the goto model; and then simulate the CBMC api to verify the instrumented goto program. If there are some violations---loop invariants are not inductive, or some pointer checks fail---, it record valuation from trace generated by the back end to construct a formatted counterexample `cext`.

Counterexample
---
A counterexample `cext` record valuations of variables in the trace, and other information about the violation. The valuation we record including
1. set of live variables upon the entry of the loop.
2. the havoced value of all primitive-typed variables;
3. the havoced offset and the object size of all pointer-typed variables;
4. history values of 2 and 3.

The valuations will be used as true positive (history values) and true negative (havcoed valuation) to filter out bad invariant clause with the idea of the Daikon invariant detector in a following PR. However, in this PR we only construct the valuation but not actually use them.

Synthesizer
---
Loop invariants we synthesize are of the form
`` (in_clause || !guard) && (!guard -> pos_clause)``
where `in_clause` and `out_clause` are predicates we store in two different map, and `guard` is the loop guard. The idea is that we separately synthesize the condition `in_clause` that should hold before the execution of the loop body, and condition `pos_clause` that should hold as post-condition of the loop.

When the violation happen in the loop, we update `in_clause`. When the violation happen after the loop, we update `pos_clause`. When the invariant candidate it not inductive, we enumerate strengthening clause to make it inductive.

To be more efficient, we choose different synthesis strategy for different type of violation
* For out-of-boundary violation, we choose to use the violated predicate as the new clause, which is the WLP of the violation if the violation is only dependent on the havocing instruction. **TODO**: to make it more complete, we need to implement a WLP algorithm
* For null-pointer violation, we choose `__CPROVER_same_object(ptr, __CPROVER_loop_entry(ptr))` as the new clause. That is, the havoced pointer should points to the same object at the start of every iteration of the loop. It is a heuristic choice. This can be extended with the idea of alias analysis if needed.
* For invariant-not-preserved violation, we enumerate strengthening clauses and check that if the invariant will be inductive after strengthening (disjunction with the new clause).

The synthesizer works as follow
1. initialize the two invariant maps,
2. verify the current candidates built from the two maps,
_a. return the candidate if there is no violation
_b. synthesize a new clause to resolve the **first** violation and add it to the correct map,
3. repeat 2.

The flag `synthesize-loop-invariants` will also apply synthesized loop contracts.

qinheping committed:
Merge branch 'diffblue:develop' into loop_invariant_synthesis

qinheping committed:
Disallow enumerate equal expr between sub-exprs with different types.

qinheping committed:
Merge branch 'diffblue:develop' into loop_invariant_synthesis

tautschnig commented:
> It sounds great to me. I will create another PR to initialize the new tool `goto-synthesizer`. At the same time, let's keep pushing this PR to be merged. Once the PR for the new tool is approved, I will migrate the loop-contract synthesizer to the new tool.

Now that this has happened: can this PR directly go in the new tool? It will avoid adding a whole bunch of dependencies into goto-instrument.

qinheping committed:
Add goto-convert before verifying candidate

qinheping commented:
> > It sounds great to me. I will create another PR to initialize the new tool `goto-synthesizer`. At the same time, let's keep pushing this PR to be merged. Once the PR for the new tool is approved, I will migrate the loop-contract synthesizer to the new tool.
> 
> Now that this has happened: can this PR directly go in the new tool? It will avoid adding a whole bunch of dependencies into goto-instrument.

Agreed. I created two PRs:
#7429 for Migrating the current synthesizer interface and the expression enumerator from `goto-instrument` into `goto-synthesizer`.
#7430 that make the same changes as in this PR but in `goto-synthesizer`.
